overlord/snapstate/snapmgr.go

```diff
@@ -387,10 +399,97 @@ func (m *SnapManager) blockedTask(cand *state.Task, running []*state.Task) bool
 	return false
 }
 
+var CanAutoRefresh func(st *state.State) bool
+
+// ensureRefreshes ensures that we refresh all installed snaps periodically
+func (m *SnapManager) ensureRefreshes() error {
+	m.state.Lock()
+	defer m.state.Unlock()
+
+	// init so that empty state gets refreshed right away
+	nextRefresh := time.Now()
+	err := m.state.Get("next-auto-refresh-time", &nextRefresh)
+	if err != nil && err != state.ErrNoState {
+		return err
+	}
+
+	// see if it even makes sense to try to refresh
+	if CanAutoRefresh == nil || !CanAutoRefresh(m.state) {
+		return nil
+	}
+
+	if time.Now().Before(nextRefresh) {
+		return nil
+	}
+
+	// check that there is no change in flight already
+	for _, chg := range m.state.Changes() {
+		if chg.Kind() == "auto-refresh" && !chg.Status().Ready() {
+			// change already in motion
+			return nil
+		}
+	}
+
+	// For scheduleNextRefresh() we have two options here:
+	//  1. schedule the next refresh only if no error happened
+	//     below in the "AutoRefresh()" call
+	//  2. always schedule the next refresh regardless of errors
+	//     in "AutoRefresh()"
+	//
+	// This currently implements (2) because if there are e.g. store
+	// errors we don't want to put more burden on the store by having
+	// every client in the world retry every 10min.
+	//
+	// However (1) is better for e.g. network errors on the client
+	// side (not yet connected etc)
+	nextRefresh = scheduleNextRefresh(m.state)
+	updated, tasksets, err := AutoRefresh(m.state)
+	if err != nil {
+		return err
+	}
+
+	var msg string
+	switch len(updated) {
+	case 0:
+		// check in after some hours
+		logger.Noticef("No snaps to auto-refresh found, will check again at %v", nextRefresh)
+		return nil
+	case 1:
+		msg = fmt.Sprintf(i18n.G("Refresh snap %q"), updated[0])
+	default:
+		quoted := strutil.Quoted(updated)
+		// TRANSLATORS: the %s is a comma-separated list of quoted snap names
+		msg = fmt.Sprintf(i18n.G("Refresh snaps %s"), quoted)
+	}
+
+	// Add "schedule-next-refresh" after a successful refresh. This ensures
+	// that if the refresh takes e.g. 7h we don't hit the store again 1h
+	// later. We could remove this to keep the behaviour simpler.
+	nextRefreshTask := m.state.NewTask("schedule-next-refresh", i18n.G("Schedule next refresh"))
+	for _, ts := range tasksets {
+		nextRefreshTask.WaitAll(ts)
+	}
+	tasksets = append(tasksets, state.NewTaskSet(nextRefreshTask))
+
+	chg := m.state.NewChange("auto-refresh", msg)
+	for _, ts := range tasksets {
+		chg.AddAll(ts)
+	}
+	chg.Set("snap-names", updated)
+	chg.Set("api-data", map[string]interface{}{"snap-names": updated})
```

Comment:
Hmm.. wondering why we have both of these.. feels like the latter should be enough, probably. Not a problem for this PR if that's how it's done already elsewhere, though.